pkg/assembler/grammar.go

```diff
@@ -0,0 +1,199 @@
+package assembler
+
+import (
+	"fmt"
+	"math"
+)
+
+// Grammar and AST
+
+type CasmProgram struct {
+	Instructions []Instruction `@@`
+}
+
+type Instruction struct {
+	Core      *CoreInstruction `@@`
+	ApPlusOne bool             `( "," @"ap" "+" "+" )? ";" |`
```

Comment:
```suggestion
	ApPlusOne bool             `( "," @"ap++" )? ";" |`
```
Not sure if this makes a difference with participle's whitespace handling, but it is a bit easier to read for me.

Reply:
Yes, I initially put it like that, but participle uses the same lexer as go, meaning it detects it treats the tokens as a golang tokens. In this case, the lexer detects both tokens separately instead of together. There is no escaping this without implementing our own lexer